Libraries/LibMarkdown/Table.cpp

```diff
@@ -0,0 +1,221 @@
+    StringBuilder format_builder;
+    StringBuilder builder;
+
+    auto write_aligned = [&](const auto& text, auto width, auto alignment) {
+        size_t original_length = 0;
+        for (auto& span : text.spans())
+            original_length += span.text.length();
+        auto string = text.render_for_terminal();
+        format_builder.clear();
+        if (alignment == Alignment::Center) {
+            auto padding_length = (width - original_length) / 2;
+            builder.appendf("%*s%s%*s", padding_length, "", string.characters(), padding_length, "");
+            if ((width - original_length) % 2)
+                builder.append(' ');
+        } else {
+            format_builder.appendf("%%%s%zus", alignment == Alignment::Left ? "-" : "", width + (string.length() - original_length));
+            builder.appendf(
+                format_builder.to_string().characters(),
+                string.characters());
+        }
+    };
+
+    bool first = true;
+    for (auto& col : m_columns) {
+        if (!first)
+            builder.append('|');
+        first = false;
+        size_t width = col.relative_width * segment_width_proportion;
+        write_aligned(col.header, width, col.alignment);
+    }
+
+    builder.append("\n");
+    for (size_t i = 0; i < view_width; ++i)
+        builder.append('-');
+    builder.append("\n");
+
+    for (size_t i = 0; i < m_row_count; ++i) {
+        bool first = true;
+        for (auto& col : m_columns) {
+            if (col.rows.size() <= i)
+                continue;
+            auto& cell = col.rows[i];
+
+            if (!first)
+                builder.append('|');
+            first = false;
+
+            size_t width = col.relative_width * segment_width_proportion;
+            write_aligned(cell, width, col.alignment);
+        }
+        builder.append("\n");
+    }
+
+    return builder.to_string();
+}
+
+String Table::render_to_html() const
+{
+    StringBuilder builder;
+
+    builder.append("<table>");
+    builder.append("<thead>");
+    builder.append("<tr>");
+    for (auto& column : m_columns) {
+        builder.append("<th>");
+        builder.append(column.header.render_to_html());
+        builder.append("</th>");
+    }
+    builder.append("</tr>");
+    builder.append("</thead>");
+    builder.append("<tbody>");
+    for (size_t i = 0; i < m_row_count; ++i) {
+        builder.append("<tr>");
+        for (auto& column : m_columns) {
+            if (column.rows.size() > i) {
+                builder.append("<td>");
+                builder.append(column.rows[i].render_to_html());
+                builder.append("</td>");
+            }
+        }
+        builder.append("</tr>");
+    }
+    builder.append("</tbody>");
+    builder.append("</table>");
+
+    return builder.to_string();
+}
+
+OwnPtr<Table> Table::parse(Vector<StringView>::ConstIterator& lines)
+{
+    auto peek_it = lines;
+    auto first_line = *peek_it;
```

Comment:
I don't really get the point of using two separate iterators here.

For the cases where the markdown is malformed, yet this was clearly meant to be a table, it's fine to `return nullptr;` while leaving `lines` in the middle of the supposed-table. It's only important that you don't move `lines` until you commit to this being a table.

Reply:
`peek_it` is only used for the first two lines, which is how we decide if this is a table.
example:
|a|b|
|---|
|foo|

is not a table

but

|a|b|
|---|---|
|foo|

is a table.

Reply:
I'd treat

```
|a|b|
|---|
|foo|
```

as a (broken) table too. CommonMark doesn't specify a syntax for tables, so we're good.

Reply:
so what should it do? 
- just ignore the extra columns?
- fail right in the delimiters?
- ?

Reply:
`return nullptr;` the moment something is too broken to handle, and from there on it's somebody else's responsibility; most likely it's going to be a Paragraph with some Text.

Which reminds me, Paragraph has special code to detect if it's actually should fail parsing and defer to some other kind of block:

https://github.com/SerenityOS/serenity/blob/e3e5e57fdefa90fd682d6a8a6f6b64ce311ed93e/Libraries/LibMarkdown/Paragraph.cpp#L64-L79

Perhaps you should add some support for tables there. Or, since it's not quite clear how that would work (detecting that the line contains `|`s? but what if they're escaped or inside <code>`</code>s?), make super sure Paragraph is always the last type of Block to be tried, and remove all that mess.

Reply:
Hmmm...does it
(no whitespace)
|work|like|that|
|---|---|---|
|I|wonder|?|

(yes whitespace)
well, I can't really check for "is it a table" in two places, perhaps there should be a `Block::can_be_parsed()` which Paragraph tests with?

Either way, since the semantics of tables are defined by us, we can just make whitespace mandatory before them; it's visually a separate paragraph after all

Reply:
I'm not talking about whitespaces, even.

I'm saying that as of now, there's no guarantee in which order we ask various block types to try and see if the current line can be parsed as that block. `Paragraph` can, in principle, eat anything, as it just wraps `Text`, so it needs to either:
* always be checked last,
* explicitly reject lines that look like they belong to other types of blocks.

Right now the latter option is implemented. I'm proposing we switch to the former (guarantee we always try Paragraph last) and remove all of this logic from `Paragraph::parse()`, and have it just call `Text::parse()`.

Reply:
"always be checked last" doesn't really work out though, a paragraph eats all the lines it can (as of right now, until it reaches something that can be parsed as another block).
unless we make `Line`s instead, and join them later, perhaps; but that just reeks of doing things twice
---

foobar
#### This is a test heading which comes right after a "paragraph"

Reply:
Right, so we can both require a newline before a table and always check paragraphs last.

That would still mean we lose non-newlined lists though, that'd be quite sad.

Reply:
well, the only other solution that comes to me is the `Line` one, we'd have the cake (even if it's in pieces) and we can eat it too?

Reply:
-> 886c6d9